rllib/utils/metrics/stats.py

```diff
@@ -383,21 +438,39 @@ def merge_in_parallel(self, *others: "Stats") -> None:
         assert all(self._window == o._window for o in others)
         assert all(self._ema_coeff == o._ema_coeff for o in others)
 
-        # Extend `self`'s values by all `others`' values.
-        for o in others:
-            self.values.extend(o.values)
-
-        # Reduce over the entire values (no matter the window size!) first.
-        store_win = self._window
-        self._window = None
-        reduced_value, new_values = self._reduced_values()
-        self._window = store_win
-        # x-fold the reduced_value over the actual window size and use the resulting
-        # list as new self.values.
-        if self._reduce_method == "mean":
-            self.values = [reduced_value] * self._window
-        else:
-            self.values = new_values
+        win = self._window or float("inf")
+
+        # Take turns stepping through `self` and `*others` values, thereby moving
+        # backwards from last index to beginning and will up the resulting values list.
+        # Stop as soon as we reach the window size.
+        new_values = []
+        tmp_values = []
+        # Loop from index=-1 backward to index=start until our new_values list has
+        # at least a len of `win`.
+        for i in range(1, max(map(len, [self, *others])) + 1):
```

Comment:
What's the performance of all the metrics collections?

Reply:
Great question!
Seems ok :D (see PPO + Atari benchmark).

My thoughts on this:
* I think the only thing that's always a little expensive is to do `MetricsLogger.log_dict()`, which used to create lots of NestedDicts and then also had to loop through them. We already got rid of NestedDict, construction of which was insanely expensive.
* Here, we simply loop through the indices of some lists (random access, should be O(0)-performance), then mean-reduce over these small lists (the size of the lists is the number of learners or env runners, so not too huge).
* We might want to decrease the default setting for `metrics_num_episodes_for_smoothing` to increase performance, but I'm not sure yet. This would also "cheat" wrt older RLlib builds, b/c then it would seemingly learn faster where nothing really changed, just the way we report. So I'm leaving it at 100 for now.